Fix CSV upload where row contains invalid URL

ChrisBAshton opened the pull request:
This PR makes the 'CSV import links' feature more robust, so that if it encounters a row which has a malformed URL it will fail gracefully rather than throw the default Rails 'something went wrong' page.

I've customised the show.html.erb file a little to allow for array-like errors, separated by `<br>`.

It also includes some README improvements, as getting the app set up took a little while.

|Before|After|
|------|-----|
|![](https://user-images.githubusercontent.com/5111927/86378860-0abab500-bc82-11ea-911c-d54ed9dd56c6.png)|![Screenshot 2020-07-03 at 10 18 47](https://user-images.githubusercontent.com/5111927/86454060-a39b1000-bd16-11ea-8da8-c7fce5c442b7.png)|

Trello: https://trello.com/c/5IFlmQQQ/2039-3-improve-local-links-manager-error-handling-for-the-csv-upload-feature

ChrisBAshton committed:
Fix heading semantics

ChrisBAshton committed:
Document setup instructions and move docs into docs/

We now use govuk-docker to run the application locally.

I've moved some documentation into the docs/ folder, for
consistency, as there was an odd split of what was in the folder
and what was in the README.

I've removed the link to the Production environment, as other
app READMEs don't have this and it doesn't fit the heading of
'live examples'.

ChrisBAshton committed:
Fix 'No user' error on app initialisation

ChrisBAshton committed:
Make import script robust to invalid URLs

When a CSV contains a 'New URL' that is invalid, we currently
show the default Rails 'Something went wrong' page, which is a
poor user experience as it doesn't give any clues as to what went
wrong. It also didn't log the error in Sentry.

We now:

- Log the error in Sentry, along with detailed debug information
- Capture the error and prevent the Rails default error page
- Present a custom error to the user, giving them the exact line
  the issue happens on. If there are multiple issues, we show
  all of the errors, separated by line breaks (<br>).

Trello: https://trello.com/c/5IFlmQQQ/2039-3-improve-local-links-manager-error-handling-for-the-csv-upload-feature

ChrisBAshton committed:
Refactor create_csv function

Simpler, and scales to cope with whatever number of links is
passed. Also makes it clearer the relationship with the 'new_url'
variable set in the context of each test.

ChrisBAshton committed:
Include the broken URL in the validator error message

This will make the Sentry error more informative.

thomasleese commented on db/seeds.rb:
```diff
@@ -5,3 +5,5 @@
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
+
+User.create!(name: "Foo", email: "foo@gov.uk")
```
I think it's better to use `find_or_create!` here so that the seed task copes with existing data and can be run multiple times.

ChrisBAshton committed:
Fix Sentry error reporting

ActiveRecord::RecordInvalid is ignored from Sentry:
https://github.com/alphagov/govuk_app_config/blob/9dd5e51a1b3c7401dd3aae6ff29a64ffa9e7cd76/lib/govuk_app_config/configure.rb\#L30

This means when the error is encountered, it is not logged in
Sentry, even if we explicitly call GovukError.notify (as it is
still the same exception type).

I've added a 'UrlValidationException' error type to
LocalLinksManager::Import, and now throw that exception
instead of the default ActiveRecord one, which should mean
that we see the log in Sentry.

thomasleese reviewed (approved):
This looks good to me - I've left one comment but I don't feel particularly strongly about it.

kevindew commented on lib/local-links-manager/import/links.rb:
```diff
@@ -1,31 +1,47 @@
+            GovukError.notify(
+              LocalLinksManager::Import::UrlValidationException.new(e.message),
+              extra: slugs.merge(link_id: link.id),
+            )
```
Sorry for a drive-by comment.

We shouldn't use Sentry for logging validation errors in a users input as per [rfc-87](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-087-dealing-with-errors.md#2-notifications-should-be-actionable): 

```
### 2. Notifications should be actionable

Sentry notifications should be something that requires a developer of the app to do something about it. Not just a piece of information.
```

As these create error logs that aren't actionable or resolvable and may appear at any cadence.

The approach we've used to record information where we want a record of user errors is to write to the Rails logger so that the issue is available in Kibana.

ChrisBAshton replied:
Thanks for the comment, Kevin. Sentry logging was a request on the Trello card, but in hindsight it does feel a bit odd to be logging an error the developer shouldn't do anything about.

I'll talk to Platform Health about a follow-up piece of work to swap this out for Rails logging.
